### PR TITLE
[Zurich] (671) Response templates for report admin

### PR DESCRIPTION
**Pull request by dracos**
For mysociety/FixMyStreet-Commercial#671.

### CODE REVIEW

File: db/schema_0033-response-templates.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+create table response_templates (
+    id serial not null primary key,
+    body_id int references body(id) not null,
+    title text not null,
+    text text not null,
+    created timestamp not null default ms_current_timestamp(),
```
Comment (dracos) [inline db/schema_0033-response-templates.sql]: This doesn't match the schema.

Comment (davea) [inline db/schema_0033-response-templates.sql]: I've added a fixup commit to remove `whenedited` from schema

File: perllib/FixMyStreet/App/Controller/Admin.pm [hunk 2]
```diff
@@ -824,6 +824,82 @@ sub report_edit : Path('report_edit') : Args(1) {
...
+    if ($c->req->method eq 'POST') {
+        if ($c->req->param('delete_template') eq _("Delete template")) {
+            $template->delete;
+        } else {
+            $template->title( $c->req->param('title') );
+            $template->text ( $c->req->param('text') );
+            $template->update_or_insert;
+
+            # log_edit currently has hard-coded types, do we need auditing on this?
+            # $c->forward( 'log_edit', [ $object_id, $object_type, $action_performed ] );
```
Comment (davea) [inline perllib/FixMyStreet/App/Controller/Admin.pm]: This should either be turned into a `log_edit` call or removed

File: perllib/FixMyStreet/DB/Result/Problem.pm [hunk 3]
```diff
@@ -660,6 +660,48 @@ sub body {
...
+Similar to C<body> but returns a body object, either from C<external_body>
+or from C<bodies_str>.
+
+NB: For cobrands where C<bodies_str> is a comma-separated list, this may
+return the first retrieved body at random.  (This behaviour is currently
+intended for Zurich, which uses a single body.)
+
+=cut
+
+sub body_object {
```
Comment (davea) [inline perllib/FixMyStreet/DB/Result/Problem.pm]: This doesn't ever seem to be used anywhere, and should probably have been removed when I updated the template to use `problem->response_templates`.

File: perllib/FixMyStreet/App/Controller/Admin.pm [hunk 4]
```diff
@@ -824,6 +824,82 @@ sub report_edit : Path('report_edit') : Args(1) {
...
+
+    # e.g. for admin
+
+    my $body = $c->model('DB::Body')->find($body_id)
+        or $c->detach( '/page_error_404_not_found' );
+
+    $self->templates_for_body($c, $body);
+}
+
+sub template_edit : Path('templates') : Args(2) {
```
Comment (davea) [inline perllib/FixMyStreet/App/Controller/Admin.pm]: Might be worth putting the `unless $c->cobrand->moniker eq 'zurich'` check on this and `templates_view` too, given that the HTML templates don't exist anywhere except the Zürich cobrand.

**Comment by davea** [comment]
A few fixup commits to address some little bits, but I think this is looking OK.

**Comment by davea** [comment]
Merged in 419b78c2c04fda6d886759cba3f10a2fd8f037f3.

**Pull request closed without merging**